Support OpenAPI 3.1.0 JsonSchemas

john-shaffer opened the pull request:
When using routes-from, the OpenAPI parser is giving me JsonSchema objects rather than the types seen in navi like ObjectSchema and StringSchema. This causes an error when there is no matching dispatch value for `spec`.

 It seems like the only reliable way to tell the "type" of the JsonSchema is to call .getTypes. This also works for ObjectSchema etc., so I changed the spec multimethod to dispatch on the result of .getTypes instead of on the class. This makes navi support any subclasses of Schema, as long as the types can be understood.

UUIDs are distinguished from normal Strings via .getFormat.

Included tests for Schema objects.

lispyclouds commented:
Thanks a lot for the PR! I'll try to have a look soon!

lispyclouds commented on src/navi/core.clj:
```diff
@@ -64,36 +57,35 @@
-  StringSchema
-  [_]
-  string?)
+  #{"string"}
```
what's the rationale for matching on a set instead of actual values like `"string"` or `"integer"`?

john-shaffer replied:
Schema objects can have [multiple types](https://json-schema.org/understanding-json-schema/reference/type), so this was the most direct way to adapt the existing code without making too many changes. It would probably work fine to detect the case of multiple types and join the specs with a malli `:or` form. Then the multimethod could dispatch on the individual scalar values. Would you prefer that approach?

john-shaffer commented:
After some more research, I believe that the problem comes from using the 3.1.0 version of OpenAPI. It seems that is what causes schemas to be parsed into JsonSchema objects rather than the older classes like StringSchema.

I added another commit that restores spec to the previous behavior of dispatching on a single type and adds a new fn, `schema->spec`. `schema->spec` can handle multi-typed schemas by joining them with a malli `[:or ,,,]`. This turns out to be important for OpenAPI 3.1.0 because `nullable: true` has been replaced with e.g. `type: ["string", "null"]`. I added a spec method for "null" to support this.

lispyclouds commented:
Looks good! Could you also update the swagger parser java dep to latest and verify the behaviour is still the same? We should be good to merge after that I think.

john-shaffer commented:
Sure. Upgraded swagger-parser to 2.1.22. Tests are green and real-world behavior is still great.

lispyclouds commented:
Looks good! Thanks!
